spec/text-editor-spec.coffee

```diff
@@ -5924,3 +5927,58 @@ describe "TextEditor", ->
   describe "::getElement", ->
     it "returns an element", ->
       expect(editor.getElement() instanceof HTMLElement).toBe(true)
+
+  describe "verticalScrollMargin", ->
+    it "scrolls the window down before reaching the bottom with a given margin", ->
+      # give plenty of text to allow for scrolling
+      editor.setText('a\n'.repeat(50))
+
+      # make sure editor element has valid dimensions
+      element = editor.getElement()
+      element.style.width  = '200px'
+      element.style.height = '120px'
+      jasmine.attachToDOM(element)
+
+      component = element.getComponent()
+      # for a height of 120px and a line height of 17px, this will be 7 rows
+      rowCount = component.getLastVisibleRow()
+
+      for scrollMargin in [0..4]
+        editor.update({verticalScrollMargin: scrollMargin})
+
+        # margin is clamped so that it is no larger than half the screen
+        maxScrollMargin = Math.floor(((component.getScrollContainerClientHeight() / editor.getLineHeightInPixels()) - 1) / 2)
```

Comment:
Hi Lee,
This clamping is actually for the right hand side of the expect(...).toBe test below, i.e. it's needed for computing the expected value for the feature. The feature itself implements this behavior, so this test ensure that it matches that behavior.